CIPs/cip-120.md

```diff
@@ -0,0 +1,109 @@
+---
+cip: 120
+title: Multiple Pubsub Topics
+author: Joel Thorstensson <joel@3box.io>
+discussions-to: https://forum.ceramic.network/t/cip-120-multiple-pubsub-topics/
+status: Draft
+category: Standards
+type: Core
+created: 2022-12-13
+edited: 2022-12-13
+---
+
+## Simple Summary
+<!--Provide a simplified and layman-accessible explanation of the CIP.-->
+Make the Ceramic protocol use multiple pubsub topics to gossip updates.
+
+
+## Abstract
+<!--A short (~200 word) description of the technical issue being addressed.-->
+
+This CIP introduces the idea of separating into multiple pubsub topics based on a separator field in a streams genesis event, creating a unique pubsub topic for each value of this separator key. For streams that doesn't specify this field, the current pubsub topic is used for backwards compatibitliy.
+
+
+## Motivation
+<!--Motivation is critical for CIPs that want to change the Ceramic protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the CIP solves. CIP submissions without sufficient motivation may be rejected outright.-->
+Currently Ceramic uses one pubsub topic to propagate *update*, *query*, and *response* messages in the network. As the network grows the scalability of this approach will be limited. Separating messages into multiple topics doesn't neccessarily solve this problem entirely. However, it allows developers of less popular streams to not share the burden message propagation for more scaled applications.
+
+
+## Specification
+<!--The technical specification should describe the syntax and semantics of any new feature.-->
+
+This CIP proposes a way to create a pubsub topic for any given stream given a *separator key* and a *separator value*. The *separator key* is defined by the `sep` field in the genesis event header and must be a string. The *separator value* is defined as the value of the field in the genesis header corresponding to the value of `sep`. The *separator value* is expected to be a byte string.
+
+All pubsub messages for streams that include a value for the `sep` field in the genesis header MUST use the following topic:
+
+```markdown
+/ceramic/<network-name>/<separator-key>/<separator-value>
+```
+
+- `network-name` - the name of the Ceramic network used
+- `separator-key` - the value of the `sep` separator key field
+- `separator-value` - the value of the field specified by the `sep` field, encoded as `base36`
+
+### Example
+
+If a stream has the following genesis header,
+
+```javascript
+{
+  header: {
+    sep: 'model',
+    model: <bytes-k2t6wyfsu4pg1hnttc39koxlw0pdn1kemzrn33cdfhrbooozvo7wb14sa1uqwi>
```

Comment:
What are we trying to convey with this `<bytes->` syntax?

Reply:
That this is in fact not JSON, but we have the bytes that represents the streamid above.

Reply:
we could make it look not JSON by putting in javascript
```
base36.decode("k2t6wyfsu4pg1hnttc39koxlw0pdn1kemzrn33cdfhrbooozvo7wb14sa1uqwi")
```

Reply:
```suggestion
    model: toBytes('k2t6wyfsu4pg1hnttc39koxlw0pdn1kemzrn33cdfhrbooozvo7wb14sa1uqwi', 'base36'))
```
makes sense!